Task_1_4_1/src/main/java/RecordBook.java

```diff
@@ -0,0 +1,118 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.OptionalDouble;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class RecordBook {
+
+    private final List<Semester> semesterList = new ArrayList<>();
+    private Subject diplomaWork;
+
+    /**
+     * Добавить семестр с оценками.
+     *
+     * @param semester добавляемый семестр.
+     */
+    public void addSemester(Semester semester) {
+        semesterList.add(semester);
+    }
+
+    /**
+     * Получить объект последнего семестра.
+     *
+     * @return объект последнего семестра, в котором можно изменять оценки.
+     */
+    public Semester getLastSemester() {
+        if (semesterList.isEmpty()) {
+            return null;
+        }
+
+        return semesterList.get(semesterList.size() - 1);
+    }
+
+    /**
+     * Установить оценку для дипломной работы.
+     *
+     * @param subject дипломная работа.
+     */
+    public void setDiplomaWork(Subject subject) {
+        diplomaWork = subject;
+    }
+
+    private Stream<Subject> getAllSubjectsStream() {
+        Stream<Subject> subjectStream = Stream.empty();
+        for (var semester : semesterList) {
+            subjectStream = Stream.concat(subjectStream, semester.getSubjectList().stream());
+        }
+        return subjectStream;
+    }
+
+    /**
+     * Получить среднюю оценку за все предметы всех семестров.
+     *
+     * @return средняя оценка.
+     */
+    public Double getAverageMark() {
+        OptionalDouble result = getAllSubjectsStream()
+                .mapToInt(i -> i.mark().value)
+                .average();
+
+        if (result.isPresent()) {
+            return result.getAsDouble();
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Возможно ли получить красный диплом.
+     *
+     * @return возможно ли?
+     */
+    public boolean isRedDiploma() {
+        if (semesterList.isEmpty() || diplomaWork == null) {
+            return false;
+        }
+
+        var lastMarkMap = getAllSubjectsStream()
+                .collect(Collectors.toMap(
+                        Subject::title,
+                        Subject::mark,
+                        (existing, replacement) -> replacement
+                ));
+
+        int fourCount = 0;
+        int fiveCount = 0;
+
+        for (var subject : lastMarkMap.entrySet()) {
+            switch (subject.getValue()) {
+                case FOUR:
+                    fourCount++;
+                    break;
+                case FIVE:
+                    fiveCount++;
+                    break;
+                default:
+                    return false;
+            }
+        }
```

Comment:
Можно так посчитать оценки с помощью стрима. Можно даже сделать чейнинг с предыдущим стримом
```suggestion
     Map<Mark, Long> count = lastMarkMap.values().stream()
            .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));

        long fourCount = count.getOrDefault(Mark.FOUR, 0L);
        long fiveCount = count.getOrDefault(Mark.FIVE, 0L);
```

Reply:
Правда тогда нужно ещё условие на отсутствие троек и двоек обработать... Так что может мой вариант и не лучше